### ISSUE TITLE
Transaction log

### ISSUE THREAD

**Issue by smspillaz** [flatpak/flatpak#1026]
Following on from #1011 , we'd also like to ensure that we don't auto-install any flatpaks in our updater that the user has already uninstalled. In order to do this, we'd need a log from flatpak about the transactions that it has performed.

My thinking about this was that there can be a log file that lives in the installation directory which specifies something like:

    timestamp event ref

Where `event` might be `install`, `uninstall`, `update`, `pull` etc.

Any thoughts on that?

**Comment by smspillaz** [comment]
CC @cosimoc @pwithnall

**Comment by pwithnall** [comment]
Each flatpak transaction can deal with multiple refs and multiple operations (for example, installing one ref might require an update of the runtime it uses). Do we want to group into transactions? I think an extensible format (JSON?) would be more appropriate than a new, custom one.

Each transaction should probably also specify the checksums of the refs it concerns. For example, a pull transaction would give the new and old checksums of the ref. What other use cases would there be for such a transaction log? Perhaps a `flatpak revert transaction $transaction_id` command? Is there a need for that? Regardless, I think it would be best to collect enough information in the transaction log that it’s possible to revert any transaction — that’s a fairly good benchmark for completeness of the logged information.

Do we intend the transaction log to be used for auditing purposes at all? If so, a load of other things have to be considered, like making it tamper-proof and including information about the user who initiated a transaction and how it was initiated. I don’t think that’s a use case we want to cover (it would be better served by writing to the system audit log), but we should be explicit about considering that.

**Comment by cosimoc** [comment]
One idea that about the format for this I had yesterday is to commit the transaction log as metadata in Flatpak's OSTree repository itself, as opposed to a separate file. That way we would not have to worry about keeping the lifecycle of the two in sync, and we'd have the same atomicity guarantees that we have elsewhere. It would also simplify how to treat user/system installations, since each repository would have its own log.

**Comment by smspillaz** [comment]
@cosimoc By "metadata" do you mean per-commit metadata? My understanding is that the flatpak repo is intended to be a local clone of the remote and doesn't know anything about installations. What ref would we attach this metadata to?

**Comment by smspillaz** [comment]
> Each flatpak transaction can deal with multiple refs and multiple operations (for example, installing one ref might require an update of the runtime it uses). Do we want to group into transactions? I think an extensible format (JSON?) would be more appropriate than a new, custom one.

I'm not tied to formats, though in my view simpler is better.

> Each transaction should probably also specify the checksums of the refs it concerns. For example, a pull transaction would give the new and old checksums of the ref. What other use cases would there be for such a transaction log? Perhaps a flatpak revert transaction $transaction_id command? Is there a need for that? Regardless, I think it would be best to collect enough information in the transaction log that it’s possible to revert any transaction — that’s a fairly good benchmark for completeness of the logged information.

I think this is definitely a good benchmark, though I'm mindful of feature creep here. Thinking narrowly, the reason why we wanted this was just so that our updater could check *if* a transaction was done.

**Comment by smspillaz** [comment]
Okay, I think what I'll do for now is put the transaction log into the installation directory as an array of JSON objects with two mandatory fields, `version` and `event`. Other fields can be added in the future if we bump the version - it would be up to the consumer of the log file to read the version field and determine what should be available accordingly.

**Comment by pwithnall** [comment]
Sounds good to me. Please be sure to document the JSON format somewhere (ideally by providing a [JSON Schema](http://json-schema.org/) for it, although that’s a bit of a big ask if you’re unfamiliar with JSON Schema; if so, I could do one at some point).

**Comment by smspillaz** [comment]
> Sounds good to me. Please be sure to document the JSON format somewhere (ideally by providing a JSON Schema for it, although that’s a bit of a big ask if you’re unfamiliar with JSON Schema; if so, I could do one at some point).

Happy to do this (it doesn't look *too* difficult). Where would we want to store the scheme within the source?

**Comment by pwithnall** [comment]
> Happy to do this (it doesn't look *too* difficult). Where would we want to store the scheme within the source?

I’d probably stick it in `common/`, where the rest of the code handling the transaction log is.

**Comment by cosimoc** [comment]
>@cosimoc By "metadata" do you mean per-commit metadata? My understanding is that the flatpak repo is intended to be a local clone of the remote and doesn't know anything about installations. What ref would we attach this metadata to?

Yeah I meant per-commit metadata, which I think could be added to the commit after the fact. But nevermind if it sounds too complicated :-)

**Comment by cgwalters** [comment]
libostree does support detached metadata on commits, though that's more intended for use by GPG signatures from remote repos.

We could definitely extend libostree to have an explicit "local" metadata object type associated with commits that e.g. wouldn't be copied by `pull-local`.

Might be useful if the systemd journal had some concept of "pinned" messages that were somehow bound to a lifecycle determined by the logging app; so for example one could log a message related to e.g. an ostree ref and only have it go away when the ref does.  I can't imagine it'd be trivial work but there'd be a lot of advantages to this model.

**Comment by cosimoc** [comment]
Unfortunately we can’t rely on the journal being around, as we set it to only store per-boot entries (I.e. it does not persist) on Endless.

**Comment by alexlarsson** [comment]
closing this as per the PR

**Comment by Mikaela** [comment]
If I understood correctly, this was closed in favour of PR https://github.com/flatpak/flatpak/pull/1030 which also got closed without this being implemented? If so, could this issue be reopened?

I would also be missing for something like `snap changes` in flatpak.

```
% snap changes --help   
Usage:
  snap changes [changes-OPTIONS] [<snap>]

The changes command displays a summary of system changes performed recently.

[changes command options]
          --abs-time  Display absolute times (in RFC 3339 format). Otherwise, display relative times up to 60 days, then YYYY-MM-DD.

% snap changes
ID   Status  Spawn                    Ready                    Summary
704  Done    yesterday at 20:30 EEST  yesterday at 20:31 EEST  Refresh snap "core"
705  Done    yesterday at 20:39 EEST  yesterday at 20:39 EEST  Refresh all snaps: no updates
706  Done    yesterday at 22:09 EEST  yesterday at 22:09 EEST  Refresh all snaps: no updates
707  Done    yesterday at 22:19 EEST  yesterday at 22:19 EEST  Refresh all snaps: no updates
708  Done    yesterday at 23:15 EEST  yesterday at 23:15 EEST  Refresh snap "zeronet"
709  Done    today at 00:40 EEST      today at 00:40 EEST      Refresh all snaps: no updates
710  Done    today at 15:31 EEST      today at 15:31 EEST      Refresh all snaps: no updates
711  Done    today at 15:33 EEST      today at 15:33 EEST      Refresh all snaps: no updates
712  Done    today at 15:38 EEST      today at 15:38 EEST      Refresh all snaps: no updates
713  Done    today at 15:41 EEST      today at 15:41 EEST      Refresh all snaps: no updates
714  Done    today at 18:51 EEST      today at 18:51 EEST      Refresh all snaps: no updates                                                     
715  Done    today at 18:51 EEST      today at 18:51 EEST      Refresh all snaps: no updates                                                     
716  Done    today at 19:37 EEST      today at 19:37 EEST      Refresh all snaps: no updates                                                     
```

**Comment by pwithnall** [comment]
> If I understood correctly, this was closed in favour of PR #1030 which also got closed without this being implemented? If so, could this issue be reopened?

Yes, this was closed because PR #1030 was closed. The PR was closed because we [found a better way to implement what we needed](https://github.com/flatpak/flatpak/pull/1030#issuecomment-332867420) to implement (installing flatpaks with OS upgrades in Endless OS).

I guess it’s up to @alexlarsson to reopen this issue if your use case for a transaction log is compelling.

**Comment by Mikaela** [comment]
I don't know how compelling my user case is, I am just thinking about home/family devices with multiple people who may be updating them and having updates in crontab just in case and I wish I had some way to know what has been done by other users than me or cron or just something to refresh my memory.
